lib/puppetserver/ca/x509_loader.rb

```diff
@@ -124,6 +124,15 @@ def load_crls(chain_path)
         return crls
       end
 
+      # Replace the CRL for the signing cert of this loader
+      #
+      # @param new_crl [OpenSSL::X509::CRL]
+      # @return [void]
+      def crl=(new_crl)
+        @crl = new_crl
+        @crls = [new_crl] + @crls.reject {|c| c.issuer == new_crl.issuer }
```

Comment:
Not sure if this applies but IIRC `OpenSSL::X509::Name` doesn't implement comparable until ruby 2.4, so this comparison may not work they way you want. I handled this in puppet by calling `X509#to_utf8`, however, I had to monkey patch that for older ruby versions 🤦‍♂

Reply:
Hmm. I cribbed this from up in the `find_leaf_crl` method where we compare `crl.issuer` against `cert.subject`, which is also a `OpenSSL::X509::Name`:

https://github.com/puppetlabs/puppetserver-ca-cli/blob/1.3.2/lib/puppetserver/ca/x509_loader.rb#L39-L41

Given the spec tests run under Ruby 2.3, it seems like some comparison logic must be in place.

Reply:
Looks like `OpenSSL::X509::Name` does include the `Comparable` module in Ruby 2.3 and also implements equality methods:

https://github.com/ruby/ruby/blob/v2_3_0/ext/openssl/ossl_x509name.c#L454

Reply:
Gah, sorry, I'm misremembering. It's `OpenSSL::X509::Certificate` that doesn't implement `#==`:

```
irb(main):004:0> left = OpenSSL::X509::Certificate.new(File.read('./spec/fixtures/ssl/ca.pem'))
=> #<OpenSSL::X509::Certificate: subject=#<OpenSSL::X509::Name:0x00007ff5599376b8>, issuer=#<OpenSSL::X509::Name:0x00007ff559937690>, serial=#<OpenSSL::BN:0x00007ff559937640>, not_before=1970-01-01 00:00:00 UTC, not_after=2029-04-19 22:31:22 UTC>
irb(main):005:0> right  = OpenSSL::X509::Certificate.new(File.read('./spec/fixtures/ssl/ca.pem'))
=> #<OpenSSL::X509::Certificate: subject=#<OpenSSL::X509::Name:0x00007ff559926660>, issuer=#<OpenSSL::X509::Name:0x00007ff559926728>, serial=#<OpenSSL::BN:0x00007ff559926818>, not_before=1970-01-01 00:00:00 UTC, not_after=2029-04-19 22:31:22 UTC>
irb(main):006:0> left == right
=> false
```

Reply:
Ah, nice to know about certificates. Looks like Ruby calls through to `X509_NAME_cmp` when comparing `OpenSSL::X509::Name` which does a comparison of the bytes in the canonical encoding of the Name:

https://github.com/openssl/openssl/blob/OpenSSL_1_1_1a/crypto/x509/x509_cmp.c#L179